src/compact-denylist-format.md

```diff
@@ -0,0 +1,504 @@
+/mime/*
+!/mime/text/plain
+!/ipns/my.domain
+/ipns/my.domain
+```
+
+In this example, `/ipns/my.domain` stays blocked because the deny rule happens
+AFTER the allow one.
+
+:::note
+
+Implementations MAY reuse denylist format for `.allow` files, where  everything
+is blocked by default, and only matching items are allowed.
+
+:::
+
+#### Hint list
+
+A hint list is an optional space-separated list of hints associated with specific block items in the form:
+
+```
+[block_item] hintA:v1 hintB:v2 hintC:v3
+```
+
+Block items and hints are separated by one or more consecutive instances of
+the "space" character.
+
+### Test fixtures
+
+Denylist parsing and correct behaviour can be tested using the
+[test.deny](https://github.com/ipfs-shipyard/nopfs/blob/master/tester/test.deny)
+denylist, which provides example rules and describes the expected behaviour in
+detail.
+
+In particular, a reference [Blocker implementation validator](https://github.com/ipfs-shipyard/nopfs/tree/master/tester) is provided in Go, and can be adapted to other languages if needed.
+
+### Security
+
+This proposal takes into account security:
+
+- Denylist headers and line-length limits are well specified to avoid malformed lists to cause things like large memory usage while parsing.
+- Supported type of blocks have been thought out to avoid amplified consumption of resources or side effects (i.e. downloading of additional dag-blocks) during the implementation.
+- Paths are sanitized and follow the same encoding rules as URLs (RFC 3986), so that existing and safe parsing can be done with regular tooling.
+- Official and custom-hint systems allow the introduction of additional features that can co-exist with the specified format without needing to be supported.
+
+### Privacy and User Control
+
+The main aspect regarding privacy in the scope of this specification has to do
+with supporting the use of double-hashing in block items.
+
+Double-hashing is particularly useful when the denylist is meant to be shared. Double-hashing:
+
+- Prevents readers of the denylist to know what the original content-address
+  of the block item is, and therefore avoids making the denylist a directory
+  of *bad* content. This is particularly useful for harmful content, where
+  solely accessing it is bad.
+- Double-hashing does not exclude adding additional context via comments or hints
+- The presence of a single double-hashed block item makes necessary that the implementation hashes every CID and CID+path that needs to be checked, which has a performance impact.
+- In general, it is good that users can inspect the nature of the content blocked if they wish to, so we recommend not using double-hashing by default as it helps transparency (i.e. blocking due to copyright claims).
+
+## Implementations
+
+- [NoPFS](https://github.com/ipfs-shipyard/nopfs): An implementation of IPIP-383 which add supports for content blocking to the go-ipfs stack and particularly to Kubo.
```

Comment:
:thought_balloon:  TODO: update once/if we know this ships built-on  in Kubo 0.24